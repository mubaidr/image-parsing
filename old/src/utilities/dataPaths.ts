import os from 'os'
import path from 'path'

const isDev: boolean = process.env.NODE_ENV === 'development'
const tmp: string = isDev ? path.resolve('.tmp') : os.tmpdir()
const home: string = path.join(os.homedir(), 'desktop')

const dataPaths = {
  design: path.resolve('__tests__', '_data', 'design.qr.svg'),
  designBarcode: path.resolve('__tests__', '_data', 'design.svg'),
  home,
  images: path.resolve('__tests__', '_data', 'images-qrcode'),
  imagesBarcode: path.resolve('__tests__', '_data', 'images-barcode'),
  key: path.resolve('__tests__', '_data', 'key.xlsx'),
  keyImage: path.resolve('__tests__', '_data', 'key.jpg'),
  questionsModel: isDev
    ? path.resolve('src/data/questions-model.json')
    : path.resolve('dist/data/questions-model.json'),
  result: path.resolve('__tests__', '_data', 'result.xlsx'),
  resultCompiled: path.resolve('__tests__', '_data', 'resultCompiled.xlsx'),
  root: path.resolve('.'),
  temp: tmp,
  tmp,
}

export { dataPaths }
