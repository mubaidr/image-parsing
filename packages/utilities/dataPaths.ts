import os from 'os';
import path from 'path';

const tmp = os.tmpdir();

export const DataPaths = {
  design: path.resolve('tests', '_data', 'design.qr.svg'),
  designBarcode: path.resolve('tests', '_data', 'design.svg'),
  home: os.homedir(),
  images: path.resolve('tests', '_data', 'images-qrcode'),
  imagesBarcode: path.resolve('tests', '_data', 'images-barcode'),
  key: path.resolve('tests', '_data', 'key.xlsx'),
  keyImage: path.resolve('tests', '_data', 'key.jpg'),
  result: path.resolve('tests', '_data', 'result.xlsx'),
  resultCompiled: path.resolve('tests', '_data', 'resultCompiled.xlsx'),
  root: path.resolve('.'),
  temp: tmp,
  tmp,
};
