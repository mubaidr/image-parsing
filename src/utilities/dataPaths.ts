import os from 'os'
import path from 'path'

const isDev: boolean = process.env.NODE_ENV === 'development'

const tmp: string = isDev ? path.resolve('.tmp') : os.tmpdir()
const home: string = path.join(os.homedir(), 'desktop')

const dataPaths = {
  design: path.resolve('_test_data', 'design.qr.svg'),
  designBarcode: path.resolve('_test_data', 'design.svg'),
  home,
  images: path.resolve('_test_data', 'images-qrcode'),
  imagesBarcode: path.resolve('_test_data', 'images-barcode'),
  key: path.resolve('_test_data', 'key.xlsx'),
  keyImage: path.resolve('_test_data', 'key.jpg'),
  questionsModel: path.resolve('src/data/questions-model.json'),
  result: path.resolve('_test_data', 'result.xlsx'),
  root: path.resolve('.'),
  temp: tmp,
  tmp,
}

export { dataPaths }
