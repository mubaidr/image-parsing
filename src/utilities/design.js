const cheerio = require('cheerio')
const fs = require('fs')

const OPTIONS = require('./question-options')

/**
 * Extracts position & dimensions of objects from SVG design File
 *
 * @param {String} designFilePath Path to the svg design file
 * @returns {Object} JSON object
 */
async function getDesignData(dir) {
  const designData = {
    questions: {},
  }
  const ROLL_NO_PATTERN = new RegExp(/rollnobarcode/gi)
  const QUESTION_PATTERN = new RegExp(/(q[1-9][0-9]?[ad])\b/gi)

  const $ = cheerio.load(fs.readFileSync(dir, 'utf8'))
  const svgViewBox = $('svg')[0].attribs.viewBox.split(' ')

  designData.width = Math.ceil(svgViewBox[2] - svgViewBox[0])
  designData.height = Math.ceil(svgViewBox[3] - svgViewBox[1])

  let gotRollNo = false
  let x
  let y
  let rx
  let ry
  let width
  let height
  let left
  let top

  const groups = $('g')
  for (let i = 0; i < groups.length; i += 1) {
    const group = groups[i]

    const title =
      $(group)
        .find('title')
        .first()
        .html()
        .trim()
        .toUpperCase() || ''

    if (!title) continue

    const isQuestionGroup = QUESTION_PATTERN.test(title)
    const isRollNoGroup =
      isQuestionGroup || gotRollNo ? false : ROLL_NO_PATTERN.test(title)

    if (!(isQuestionGroup || isRollNoGroup)) continue

    const transform = $(group)
      .attr('transform')
      .replace(/(translate)|\(|\)/gi, '')
      .split(',')
      .map(val => parseInt(val, 10) || 0)

    const rect = $(group)
      .find('rect')
      .first()

    left = parseInt(rect.attr('x'), 10) || 0
    top = parseInt(rect.attr('y'), 10) || 0

    rx = parseInt(rect.attr('rx'), 10) || 0
    ry = parseInt(rect.attr('ry'), 10) || 0

    x = left - rx + transform[0]
    y = top - ry + transform[1]

    width = parseInt(rect.attr('width'), 10) || 0 + rx
    height = parseInt(rect.attr('height'), 10) || 0 + ry

    if (isQuestionGroup) {
      const optionTitle = title.slice(-1)
      const questionNumber = title.slice(0, -1)

      if (!designData.questions[questionNumber]) {
        designData.questions[questionNumber] = {}
      }

      if (optionTitle === OPTIONS.A) {
        designData.questions[questionNumber].x1 = x - 1
        designData.questions[questionNumber].y1 = y - 1
      } else {
        // donot check last option to suport any number of options
        designData.questions[questionNumber].x2 = x + width + 1
        designData.questions[questionNumber].y2 = y + height + 1
      }
    } else if (isRollNoGroup) {
      designData.rollNo = { x1: x, y1: y, x2: x + width, y2: y + height }
      gotRollNo = true
    }
  }

  return designData
}

module.exports = {
  getDesignData,
}
