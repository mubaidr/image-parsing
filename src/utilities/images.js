const javascriptBarcodeReader = require('javascript-barcode-reader')
const javascriptQRReader = require('jsqr').default
const fastGlob = require('fast-glob')
const path = require('path')
const uuid = require('uuid')
const sharp = require('sharp')

const { IMAGES, NATIVE_IMAGES } = require('./valid-types')
const DATAPATHS = require('./data-paths')

const CACHE = {}

function getSharpObjectFromSource(src) {
  return sharp(src)
  // .jpeg()
  // .flatten()
}

async function convertImage(src) {
  // native supported images
  if (NATIVE_IMAGES.includes(src.split('.').pop())) return src

  // check cache
  const cached = CACHE[src]
  if (cached) return cached

  // generate random tmp url
  const url = path.join(DATAPATHS.tmp, `${uuid()}.jpg`)
  CACHE[src] = url

  // save file for preview
  await getSharpObjectFromSource(src).toFile(url)

  // returns new url
  return url
}

/**
 * Logs provided image data to .tmp folder
 * @param {sharp || String} img Sharp instance or image source
 * @param {String} name Name of file
 */
async function logImageData(src, name) {
  let img

  if (typeof src === 'string') {
    img = getSharpObjectFromSource(src)
  } else {
    img = src.clone()
  }

  return img.jpeg().toFile(path.join(DATAPATHS.tmp, `${name || uuid()}.jpg`))
}

/**
 * Return a list of valid image format files from the provided path
 *
 * @param {String} path Path to sarch for images
 * @param {Array.<String>} format Array of extensions of valid image formats
 * @returns {Array.<String>} List of file paths
 */
function getImagePaths(dir) {
  return fastGlob(`${dir}/*.{${IMAGES.join(',')}}`, { onlyFiles: true })
}

/**
 *
 *
 * @param {Object} designData A JSON Object containing information about the position, width, height of elements in svg design file (available from utiltities/getDesignData)
 * @param {String} path Path of scanned image file
 * @returns {Object} Roll Number
 */
async function getRollNoFromImage(designData, img, isBarcode) {
  const metadata = await img.metadata()
  const rollNoPos = designData.rollNo
  const ratio = metadata.width / designData.width
  const width = Math.ceil((rollNoPos.x2 - rollNoPos.x1) * ratio)
  const height = Math.ceil((rollNoPos.y2 - rollNoPos.y1) * ratio)
  const obj = {
    id: uuid(),
  }

  img
    .extract({
      left: Math.floor(rollNoPos.x1 * ratio),
      top: Math.floor(rollNoPos.y1 * ratio),
      width,
      height,
    })
    .ensureAlpha()

  // debug image
  // logImageData(img)

  const data = await img.toBuffer()

  try {
    let rollNo

    if (isBarcode) {
      rollNo = await javascriptBarcodeReader(
        { data, width, height },
        { barcode: 'code-39' }
      )
    } else {
      rollNo = javascriptQRReader(data, width, height, {
        inversionAttempts: 'dontInvert',
      }).data
    }

    obj.rollNo = rollNo
    obj.hasValidRollNo = !!obj.rollNo
  } catch (err) {
    // console.error(err)

    obj.rollNo = null
    obj.hasValidRollNo = false
  }

  return obj
}

module.exports = {
  convertImage,
  getImagePaths,
  logImageData,
  getRollNoFromImage,
  getSharpObjectFromSource,
}
