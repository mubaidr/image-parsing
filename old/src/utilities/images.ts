import fastGlob from 'fast-glob'
import path from 'path'
import sharp, { Sharp } from 'sharp'
import uuid from 'uuid'

import ImageNativeTypesEnum from './@enums/ImageNativeTypesEnum'
import ImageTypesEnum from './@enums/ImageTypesEnum'
import { cache } from './cache'
import { dataPaths } from './dataPaths'

const getSharpObjectFromSource = (src: string): Sharp => {
  return sharp(src)
    .raw()
    .flatten()
}

const convertImage = async (src: string): Promise<string> => {
  if (!src) {
    throw new Error('Invalid source provided')
  }

  const ext = src.split('.').pop()

  // native supported images
  if (ext && ext in ImageNativeTypesEnum) {
    return src
  }

  // check cache
  const cached = cache.get(src)
  if (cached) {
    return cached
  }

  // generate random tmp url
  const url = path.join(dataPaths.tmp, `${uuid()}.jpg`)
  cache.set(src, url)

  // save file for preview
  await getSharpObjectFromSource(src)
    .jpeg()
    .toFile(url)

  // returns new url
  return url
}

const logImageData = async (
  src: string | Sharp,
  name?: string
): Promise<string> => {
  let img: Sharp
  const target = path.join(dataPaths.tmp, `${name || uuid()}.jpg`)

  if (typeof src === 'string') {
    img = getSharpObjectFromSource(src)
  } else {
    img = src.clone()
  }

  await img.jpeg().toFile(target)

  return target
}

const getImagePaths = async (dir: string): Promise<string[]> => {
  const loc = dir.replace(/\\/g, '/')
  const exts = Object.keys(ImageTypesEnum)
  // avoid deep directory scan in test env
  const glob =
    process.env.NODE_ENV === 'test'
      ? `${loc}/*.{${exts}}`.replace('//', '/')
      : `${loc}/**/*.{${exts}}`.replace('//', '/')

  return fastGlob(glob, {
    absolute: true,
    onlyFiles: true,
  })
}

export { convertImage, getImagePaths, logImageData, getSharpObjectFromSource }
