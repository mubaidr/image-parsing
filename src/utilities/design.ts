import { parse } from 'fast-xml-parser'
import { readFileSync } from 'fs'
import { DesignData, ItemInfo } from './@classes/WorkerManager'
import RegExpPatterns from './@enums/RegExpPatterns'
import { dataPaths } from './dataPaths'

function getDesignPathByID(id: string) {
  //TODO: implement internal store for designs with import/export functionality
  console.log(id)

  return dataPaths.designBarcode
}

export async function getDesignData(id: string): Promise<DesignData> {
  const filePath = getDesignPathByID(id)

  const { svg } = parse(readFileSync(filePath).toString(), {
    attributeNamePrefix: '',
    ignoreAttributes: false,
    parseNodeValue: true,
    parseAttributeValue: true,
    allowBooleanAttributes: true,
  })

  // get width, height from viewbox
  const [x1, y1, x2, y2] = svg.viewBox
    .split(' ')
    .map((i: string) => parseInt(i, 10))
  const svgWidth = x2 - x1
  const svgHeight = y2 - y1

  // prepare pattern matching reg expressions
  const PATTERN_OPTION = new RegExp(RegExpPatterns.OPTION, 'i')
  const PATTERN_BARCODE = new RegExp(RegExpPatterns.BARCODE, 'i')
  const PATTERN_QRCODE = new RegExp(RegExpPatterns.QRCODE, 'i')

  // for export
  let code: ItemInfo = { x: 0, y: 0, width: 0, height: 0 }
  const questions: {
    [key: string]: ItemInfo
  } = {}

  svg.g.forEach((group: any) => {
    const { title, transform, rect } = group

    if (!rect) return

    const { rx, ry } = rect
    let { x, y, width, height } = rect
    let xTransform = 0
    let yTransform = 0

    if (transform) {
      ;[xTransform, yTransform] = transform
        .replace(/(translate)|\(|\)/gi, '')
        .split(',')
        .map((val: string) => parseInt(val, 10) || 0)
    }

    x = Math.floor(x - rx + xTransform - 3)
    y = Math.floor(y - ry + yTransform - 3)
    width = Math.ceil(width + rx + 6)
    height = Math.ceil(height + ry + 6)

    if (PATTERN_OPTION.test(title)) {
      const questionNumber: string = title.slice(0, -1)

      if (!questions[questionNumber]) {
        questions[questionNumber] = {
          x: svgWidth,
          y: svgHeight,
          width: 0,
          height: 0,
        }
      }

      questions[questionNumber].width += width

      if (questions[questionNumber].x > x) {
        questions[questionNumber].x = x
      }

      if (questions[questionNumber].y > y) {
        questions[questionNumber].y = y
      }

      if (questions[questionNumber].width < width) {
        questions[questionNumber].width = width
      }

      if (questions[questionNumber].height < height) {
        questions[questionNumber].height = height
      }
    } else if (PATTERN_BARCODE.test(title) || PATTERN_QRCODE.test(title)) {
      code = { x, y, width, height }
    }
  })

  return {
    code,
    questions,
    width: svgWidth,
    height: svgHeight,
  }
}
