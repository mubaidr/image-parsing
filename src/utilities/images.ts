import fastGlob from 'fast-glob'
import NodeCache from 'node-cache'
import path from 'path'
import sharp, { Sharp } from 'sharp'
import { v4 as uuid4 } from 'uuid'
import { dataPaths } from './dataPaths'

const myCache = new NodeCache()

export enum IMAGE_NATIVE_TYPES {
  'bmp' = 'bmp',
  'gif' = 'gif',
  'jfif' = 'jfif',
  'jpe' = 'jpe',
  'jpeg' = 'jpeg',
  'jpg' = 'jpg',
  'png' = 'png',
  'svg' = 'svg',
  'webp' = 'webp',
}

export enum IMAGE_TYPES {
  'bmp' = 'bmp',
  'dib' = 'dib',
  'gif' = 'gif',
  'jfif' = 'jfif',
  'jpe' = 'jpe',
  'jpeg' = 'jpeg',
  'jpg' = 'jpg',
  'png' = 'png',
  'svg' = 'svg',
  'tif' = 'tif',
  'tiff' = 'tiff',
  'webp' = 'webp',
}

export async function getSharpObjectFromSource(src: string): Promise<Sharp> {
  let sharpImage = sharp(src, {
    failOnError: true,
  })
  const { width, height } = await sharpImage.metadata()
  const padding = 25

  if (!width || !height) throw 'Invalid image file'

  sharpImage = sharp(
    await sharpImage
      .extract({
        left: padding,
        top: padding,
        width: width - padding * 2,
        height: height - padding * 2,
      })
      .toBuffer()
  )
    .trim(175)
    .flatten()
    .raw()

  // logImageData(sharpImage, `trimmed ${padding}`)

  return sharpImage
}

export async function convertImage(src: string): Promise<string> {
  if (!src) {
    throw new Error('Invalid source provided')
  }

  const ext = src.split('.').pop()

  // native supported images
  if (ext && ext in IMAGE_NATIVE_TYPES) {
    return src
  }

  // check cache
  const cached = myCache.get(src)
  if (cached !== undefined) {
    return cached as string
  }

  // // generate random tmp url
  const url = path.join(dataPaths.tmp, `${uuid4()}.jpg`)
  myCache.set(src, url)

  // save file for preview
  getSharpObjectFromSource(src).then((img) => {
    img.jpeg().toFile(url)
  })

  // returns new url
  return url
}

export async function logImageData(
  src: string | Sharp,
  name?: string
): Promise<string> {
  let img: Sharp
  const target = path.join(dataPaths.tmp, `${name || uuid4()}.jpg`)

  if (typeof src === 'string') {
    img = await getSharpObjectFromSource(src)
  } else {
    img = src.clone()
  }

  img.jpeg().toFile(target)

  return target
}

export function getImagePaths(dir: string): string[] {
  const loc = dir.replace(/\\/g, '/')
  const exts = Object.keys(IMAGE_TYPES).map((ext) => `.${ext}`)
  const glob = `${loc}/**/*{${exts}}`.replace('//', '/')

  return fastGlob.sync(glob, {
    absolute: true,
    onlyFiles: true,
    stats: false,
  })
}
