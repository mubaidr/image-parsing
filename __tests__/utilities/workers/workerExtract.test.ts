import { dataPaths } from '../../../src/utilities/dataPaths'
import { getDesignData } from '../../../src/utilities/design'
import { getImagePaths } from '../../../src/utilities/images'
import workerExtract from '../../../src/utilities/workers/workerExtract'

describe('workerExtract', () => {
  test('should be able to extract result just fine', async () => {
    const designData = getDesignData(dataPaths.designBarcode)
    const imagePaths = await getImagePaths(dataPaths.imagesBarcode)

    const results = await workerExtract.start(
      {
        designData,
        imagePaths,
      },
      false,
    )

    const result = results ? results[0] : {}

    expect(result).toMatchSnapshot({
      id: expect.any(String),
    })
  })
})
