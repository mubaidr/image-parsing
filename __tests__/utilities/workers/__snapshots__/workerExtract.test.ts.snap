// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workerExtract should be able to extract result just fine 1`] = `
Object {
  "answers": Object {
    "Q1": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "Q10": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "Q11": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "Q12": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "Q13": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "Q14": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "Q15": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "Q16": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "Q17": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "Q18": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "Q19": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "Q2": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "Q20": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "Q21": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "Q22": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "Q23": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "*",
    },
    "Q24": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "Q25": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "Q26": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "Q27": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "Q28": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "Q29": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "Q3": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "Q30": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "Q31": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "Q32": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "Q33": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "Q34": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "Q35": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "Q36": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "Q37": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "Q38": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "Q39": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "Q4": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "Q40": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "Q41": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "Q42": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "Q43": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "Q44": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "Q45": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "Q46": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "Q47": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "Q48": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "Q49": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "Q5": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "Q50": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "Q51": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "Q52": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "Q53": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "Q54": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "Q55": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "Q56": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "Q57": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "Q58": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "Q59": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "Q6": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "Q60": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "Q7": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "Q8": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "Q9": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
  },
  "correctCount": 0,
  "id": Any<String>,
  "imageFile": Anything,
  "incorrectCount": 0,
  "isCompiled": false,
  "isRollNoExtracted": true,
  "marks": 0,
  "post": "",
  "questionPaperType": "",
  "rollNo": "10023",
  "skippedCount": 0,
  "testCenter": "",
  "testTime": "",
  "totalMarks": 0,
  "unattemptedCount": 0,
}
`;
